src/tilerange.jl

```diff
@@ -0,0 +1,197 @@
+# AbstractTileRange Protocols
+#
+# The parent range `R` is supposed to be an array of scalar indices, including:
+#   - ranges
+#   - vectors of integers
+#   - CartesianIndices{1}
+#
+# A tile range generated from `R` is a iterator whose elements are also an array of scalar indices.
+# It's an indices-specific block array.
+#
+# Reference: https://docs.julialang.org/en/v1.5/manual/arrays/#man-supported-index-types
+#
+#
+# | scalar index      |   indices(tile)      |
+# |-------------------|-----------------     |
+# | range/vector      |   AbstractTileRange  |
+# | CartesianIndices  |   TileIndices        |
+#
+# AbstractTileStrategy serves as an adapter for `AbstractTileRange` and `TileIndices`
+
+
+abstract type AbstractTileRange{R} <: AbstractArray{R, 1} end
+abstract type AbstractTileStrategy end
+
+const Range1 = Union{AbstractUnitRange, AbstractVector}
+
+### FixedTileRange and FixedTile
+
+"""
+    FixedTileRange(r, n, [Δ=n]; keep_last=true)
+
+Construct a sliding tile along range `r` with fixed sliding stride `Δ` and tile length `n`.
+
+# Arguments
+
+- `r`: a range, `CartesianIndices` or `Vector`
+- `n::Integer`: The length of each tile. If keyword `keep_last=true`, the last tile length might be
+  less than `n`.
+- `Δ::Union{Integer, CartesianIndex{1}}=n`: The sliding stride `Δ` is defined as `first(r[n]) -
+  first(r[n-1])`. Using a stride `Δ<n` means there are overlaps between each adjacent tile. `Δ` can
+  also be `CartesianIndex{1}`.
+- `keep_last=true` (keyword): this keyword affects the cases when the last tile has few elements
+  than `n`, in which case, `true`/`false` tells `FixedTileRange` to keep/discard the last tile.
+
+# Examples
+
+```jldoctest; setup=:(using TiledIteration)
+julia> FixedTileRange(2:10, 3)
+3-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 2:4
+ 5:7
+ 8:10
+
+julia> FixedTileRange(1:10, 4)
+3-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 5:8
+ 9:10
+
+julia> FixedTileRange(1:10, 4, 2)
+4-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 3:6
+ 5:8
+ 7:10
+
+julia> FixedTileRange(1:10, 4; keep_last=false)
+2-element FixedTileRange{UnitRange{Int64},Int64,UnitRange{Int64}}:
+ 1:4
+ 5:8
+```
+
+Besides an `AbstractUnitRange`, the input range `r` can also be a `CartesianIndices{1}` or more
+generally, an `AbstractVector{<:Integer}`:
+
+```jldoctest; setup=:(using TiledIteration)
+julia> FixedTileRange(CartesianIndices((1:10, )), 4)
+3-element FixedTileRange{CartesianIndices{1,Tuple{UnitRange{Int64}}},Int64,CartesianIndices{1,Tuple{UnitRange{Int64}}}}:
+ [CartesianIndex(1,), CartesianIndex(2,), CartesianIndex(3,), CartesianIndex(4,)]
+ [CartesianIndex(5,), CartesianIndex(6,), CartesianIndex(7,), CartesianIndex(8,)]
+ [CartesianIndex(9,), CartesianIndex(10,)]
+```
```

Comment:
It's even more ugly and unstable when it comes to N-d case.

```julia
julia> TileIndices(CartesianIndices((1:20, 1:20)), (7, 7))
3×3 TileIndices{Tuple{CartesianIndices{1, Tuple{UnitRange{Int64}}}, CartesianIndices{1, Tuple{UnitRange{Int64}}}}, 2, FixedTileRange{CartesianIndices{1, Tuple{UnitRange{Int64}}}, Int64, CartesianIndices{1, Tuple{UnitRange{Int64}}}}}:
 CartesianIndex{2}[CartesianIndex(1, 1) CartesianIndex(1, 2) … CartesianIndex(1, 6) CartesianIndex(1, 7); CartesianIndex(2, 1) CartesianIndex(2, 2) … CartesianIndex(2, 6) CartesianIndex(2, 7); … ; CartesianIndex(6, 1) CartesianIndex(6, 2) … CartesianIndex(6, 6) CartesianIndex(6, 7); CartesianIndex(7, 1) CartesianIndex(7, 2) … CartesianIndex(7, 6) CartesianIndex(7, 7)]                  …  CartesianIndex{2}[CartesianIndex(1, 15) CartesianIndex(1, 16) … CartesianIndex(1, 19) CartesianIndex(1, 20); CartesianIndex(2, 15) CartesianIndex(2, 16) … CartesianIndex(2, 19) CartesianIndex(2, 20); … ; CartesianIndex(6, 15) CartesianIndex(6, 16) … CartesianIndex(6, 19) CartesianIndex(6, 20); CartesianIndex(7, 15) CartesianIndex(7, 16) … CartesianIndex(7, 19) CartesianIndex(7, 20)]
 CartesianIndex{2}[CartesianIndex(8, 1) CartesianIndex(8, 2) … CartesianIndex(8, 6) CartesianIndex(8, 7); CartesianIndex(9, 1) CartesianIndex(9, 2) … CartesianIndex(9, 6) CartesianIndex(9, 7); … ; CartesianIndex(13, 1) CartesianIndex(13, 2) … CartesianIndex(13, 6) CartesianIndex(13, 7); CartesianIndex(14, 1) CartesianIndex(14, 2) … CartesianIndex(14, 6) CartesianIndex(14, 7)]             CartesianIndex{2}[CartesianIndex(8, 15) CartesianIndex(8, 16) … CartesianIndex(8, 19) CartesianIndex(8, 20); CartesianIndex(9, 15) CartesianIndex(9, 16) … CartesianIndex(9, 19) CartesianIndex(9, 20); … ; CartesianIndex(13, 15) CartesianIndex(13, 16) … CartesianIndex(13, 19) CartesianIndex(13, 20); CartesianIndex(14, 15) CartesianIndex(14, 16) … CartesianIndex(14, 19) CartesianIndex(14, 20)]
 CartesianIndex{2}[CartesianIndex(15, 1) CartesianIndex(15, 2) … CartesianIndex(15, 6) CartesianIndex(15, 7); CartesianIndex(16, 1) CartesianIndex(16, 2) … CartesianIndex(16, 6) CartesianIndex(16, 7); … ; CartesianIndex(19, 1) CartesianIndex(19, 2) … CartesianIndex(19, 6) CartesianIndex(19, 7); CartesianIndex(20, 1) CartesianIndex(20, 2) … CartesianIndex(20, 6) CartesianIndex(20, 7)]     CartesianIndex{2}[CartesianIndex(15, 15) CartesianIndex(15, 16) … CartesianIndex(15, 19) CartesianIndex(15, 20); CartesianIndex(16, 15) CartesianIndex(16, 16) … CartesianIndex(16, 19) CartesianIndex(16, 20); … ; CartesianIndex(19, 15) CartesianIndex(19, 16) … CartesianIndex(19, 19) CartesianIndex(19, 20); CartesianIndex(20, 15) CartesianIndex(20, 16) … CartesianIndex(20, 19) CartesianIndex(20, 20)]
```

Reply:
Is there a way to display those `CartesianIndices` as 

`[CartesianIndex(1, 1):CartesianIndex(1, 7), CartesianIndex(1, 1):CartesianIndex(8, 15), CartesianIndex(1, 1):CartesianIndex(16, 20)]`

while still keep the current show behavior?

I vaguely remember that we have a `top_level` keyword for display related function, but can't find the name of it now.